biosynseq/metrics.py

```diff
@@ -0,0 +1,529 @@
+"""Sequence Metrics."""
+import functools
+from concurrent.futures import ProcessPoolExecutor
+from pathlib import Path
+from typing import List
+
+import numpy as np
+import pandas as pd
+from Bio import Align, SeqIO, SeqUtils
+from Bio.Seq import Seq
+from Bio.SeqRecord import SeqRecord
+from Bio.SeqUtils.IsoelectricPoint import IsoelectricPoint
+from scipy.spatial import distance_matrix
+from tqdm import tqdm
+
+
+def get_embed_avg(embed_path: Path) -> np.ndarray:
+    """Given a path to embeddings, return the average embedding.
+
+    Parameters
+    ----------
+    embed_path : Path
+        Path to access embeddings.
+        The embeddings could be for the training, validation, testing, or generated sequences.
+
+    Returns
+    -------
+    np.ndarray
+        Average embedding.
+    """
+    embed = np.load(embed_path)
+    embed_avg = embed.mean(axis=1)
+    return embed_avg
+
+
+def get_seqs_from_fasta(
+    fasta_path: Path, translate_to_protein: bool = False
+) -> List[SeqRecord]:
+    """Given a fasta file, obtain its DNA sequences or translated protein sequences.
+
+    Parameters
+    ----------
+    fasta_path : Path
+        Path to access the fasta sequences.
+    translate_to_protein : bool, optional
+        Whether to translate the DNA sequences to protein sequneces, by default False.
+
+    Returns
+    -------
+    List[SeqRecord]
+        DNA or protein sequences from the fasta file.
+    """
+    seqs = list(SeqIO.parse(fasta_path, "fasta"))
+    if translate_to_protein:
+        # translate DNA seqs to protein seqs; stop translation at the first in-frame stop codon
+        seqs = [seq.translate(to_stop=True) for seq in seqs]
+    return seqs
+
+
+def fasta_to_dna_seqs(fasta_path: Path) -> List[SeqRecord]:
```

Comment:
No need to have this function, simply use the above function.

Reply:
I had to write this function for the part in the main function of visualize.py to get alignment plots (lines 595-598), where the length of the DNA sequences must be clipped to the length of the embeddings before the DNA sequences can be translated to protein sequences. Clipping is required for the test and validation sequences, but not for generated sequences. This intermediate clipping step thus requires fasta_to_dna_seqs to access the DNA sequences to do the clipping. Do you have suggestions for a better way to go about this?